### PR TITLE
Add a UnitTesting test for the format-on-save feature

### PR DESCRIPTION
**Pull request by rwols**
This doesn't actually cover the transport/threads handling as we're dealing with a MockClient. It only verifies the "business logic" :)

### CODE REVIEW

File: tests/setup.py [hunk 1]
```diff
@@ -53,11 +53,17 @@ def close_test_view(view):
...
+        self.client = None
+
     def setUp(self) -> None:
-        self.view = sublime.active_window().open_file(test_file_path)
+        self.view = sublime.active_window().open_file(self.test_file_path)
         self.view.settings().set("auto_complete_selector", "text.plain")
         self.wm = windows.lookup(self.view.window())
-        self.client = MockClient(async_response=sublime_delayer(100))
+        if self.client is None:
+            self.client = MockClient(async_response=sublime_delayer(100))
```
Comment (tomv564) [inline tests/setup.py]: Can you explain why test_file_path must be configurable and why the change in client assignment?

File: tests/test_pre_save.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        super().tearDown()
+
+    def test_format_on_save(self):
+        """
+        Open the (empty) file on_pre_save.txt, put the character 'a' in it, and then save it.
+        The test language server should remove the 'a' character.
+        """
+        yield 100
+        setup = self.view.settings().set
+        setup("ensure_newline_at_eof_on_save", False)
```
Comment (tomv564) [inline tests/test_pre_save.py]: Prefer a specific name like `set_view_setting` (as `setup` might suggest something more involved)

File: tests/test_pre_save.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        setup("trim_trailing_white_space_on_save", False)
+        setup("lsp_format_on_save", True)
+        self.client.responses['textDocument/formatting'] = [{
+            "newText": "",
+            "range": {
+                "start": {"line": 0, "character": 0},
+                "end":   {"line": 0, "character": 1}
+            }
+        }]
+        run = self.view.run_command
```
Comment (tomv564) [inline tests/test_pre_save.py]: Not sure if we need to alias the occurrence of two not terribly-long calls?

File: plugin/core/test_mocks.py [hunk 4]
```diff
@@ -284,6 +285,16 @@ def send_request(self, request: Request, on_success: 'Callable', on_error: 'Call
         else:
             on_success(response)
 
+    def execute_request(self, request: Request) -> 'Any':
+        response = None
```
Comment (tomv564) [inline plugin/core/test_mocks.py]: This could be as simple as 

```python
return self.responses.get(request.method)
```

File: tests/setup.py (hunk 1 above)
Comment (rwols) [reply tests/setup.py]: The test saves the file. If the test fails, then other tests would use the content in that file.

As for the customizable MockClient, my initial thought was to make a test for the sync timeouts, but then I realized that we were dealing with mock transports anyway. I reverted that part.
